geode-wan/src/test/java/org/apache/geode/cache/wan/internal/WanCopyRegionFunctionServiceTest.java

```diff
@@ -0,0 +1,183 @@
+
+    assertThatThrownBy(() -> service.execute(secondExecution, regionName, senderId))
+        .isInstanceOf(WanCopyRegionFunctionServiceAlreadyRunningException.class);
+
+    // Let first execution finish
+    latch.countDown();
+  }
+
+  @Test
+  public void cancelRunningExecutionReturnsSuccess() {
+    String regionName = "myRegion";
+    String senderId = "mySender";
+    CountDownLatch latch = new CountDownLatch(1);
+    Callable<CliFunctionResult> firstExecution = () -> {
+      latch.await();
+      return null;
+    };
+    CompletableFuture
+        .supplyAsync(() -> {
+          try {
+            return service.execute(firstExecution, regionName, senderId);
+          } catch (Exception e) {
+            return null;
+          }
+        });
+
+    // Wait for the function to start execution
+    await().until(() -> service.getNumberOfCurrentExecutions() == 1);
+
+    // Cancel the function execution
+    boolean result = service.cancel(regionName, senderId);
+
+    assertThat(result).isEqualTo(true);
+    await().untilAsserted(() -> assertThat(service.getNumberOfCurrentExecutions()).isEqualTo(0));
+  }
+
+  @Test
+  public void cancelNotRunningExecutionReturnsError() {
+    final String regionName = "myRegion";
+    final String senderId = "mySender";
+
+    boolean result = service.cancel(regionName, senderId);
+
+    assertThat(result).isEqualTo(false);
+  }
+
+  @Test
+  public void cancelAllExecutionsWithRunningExecutionsReturnsCanceledExecutions() {
+    CountDownLatch latch = new CountDownLatch(2);
+    Callable<CliFunctionResult> firstExecution = () -> {
+      latch.await();
+      return null;
+    };
+
+    CompletableFuture
+        .supplyAsync(() -> {
+          try {
+            return service.execute(firstExecution, "myRegion", "mySender1");
+          } catch (Exception e) {
+            return null;
+          }
+        });
+
+    Callable<CliFunctionResult> secondExecution = () -> {
+      latch.await();
+      return null;
+    };
+
+    CompletableFuture
+        .supplyAsync(() -> {
+          try {
+            return service.execute(secondExecution, "myRegion", "mySender");
+          } catch (Exception e) {
+            return null;
+          }
+        });
+
+    // Wait for the functions to start execution
+    await().until(() -> service.getNumberOfCurrentExecutions() == 2);
+
+    // Cancel the function execution
+    String executionsString = service.cancelAll();
+
+    assertThat(executionsString).isEqualTo("[(myRegion,mySender1), (myRegion,mySender)]");
+    await().untilAsserted(() -> assertThat(service.getNumberOfCurrentExecutions()).isEqualTo(0));
+  }
+
+  @Test
+  public void severalExecuteWithDifferentRegionOrSenderAreAllowed() {
+    int executions = 5;
+    CountDownLatch latch = new CountDownLatch(executions);
+    for (int i = 0; i < executions; i++) {
+      Callable<CliFunctionResult> firstExecution = () -> {
+        latch.await();
```

Comment:
Since `CountDownLatch` await waits forever by default, you should change it to:
```
latch.await(GeodeAwaitility.getTimeout().getSeconds(), TimeUnit.SECONDS);
```